File: javascript/ch04/11.trim-messages.js
import {
  AIMessage,
  HumanMessage,
  SystemMessage,
  trimMessages,
} from '@langchain/core/messages';
import { ChatOpenAI } from '@langchain/openai';

const messages = [
  new SystemMessage('당신은 친절한 어시스턴트입니다.'),
  new HumanMessage('안녕하세요! 나는 민혁입니다.'),
  new AIMessage('안녕하세요!'),
  new HumanMessage('바닐라 아이스크림을 좋아해요.'),
  new AIMessage('좋네요!'),
  new HumanMessage('2 + 2는 얼마죠?'),
  new AIMessage('4입니다.'),
  new HumanMessage('고마워요.'),
  new AIMessage('천만에요!'),
  new HumanMessage('즐거운가요?'),
  new AIMessage('예!'),
];

const trimmer = trimMessages({
  maxTokens: 65,
  strategy: 'last',
  tokenCounter: new ChatOpenAI({ modelName: 'gpt-4o' }),
  includeSystem: true,
  allowPartial: false,
  startOn: 'human',
});

const trimmed = await trimmer.invoke(messages);
console.log(trimmed);

File: python/ch04/11.trim-messages.py
from langchain_core.messages import (
    SystemMessage,
    HumanMessage,
    AIMessage,
    trim_messages,
)
from langchain_openai import ChatOpenAI

# 샘플 메시지 설정
messages = [
    SystemMessage(content='당신은 친절한 어시스턴트입니다.'),
    HumanMessage(content='안녕하세요! 나는 민혁입니다.'),
    AIMessage(content='안녕하세요!'),
    HumanMessage(content='바닐라 아이스크림을 좋아해요.'),
    AIMessage(content='좋네요!'),
    HumanMessage(content='2 + 2는 얼마죠?'),
    AIMessage(content='4입니다.'),
    HumanMessage(content='고마워요.'),
    AIMessage(content='천만에요!'),
    HumanMessage(content='즐거운가요?'),
    AIMessage(content='예!'),
]

# 축약 설정
trimmer = trim_messages(
    max_tokens=65,
    strategy='last',
    token_counter=ChatOpenAI(model='gpt-4o-mini'),
    include_system=True,
    allow_partial=False,
    start_on='human',
)

# 축약 적용
trimmed = trimmer.invoke(messages)
print(trimmed)
